dwave/optimization/include/dwave-optimization/array.hpp

```diff
@@ -737,6 +738,27 @@ std::vector<ssize_t> broadcast_shape(const std::span<const ssize_t> lhs,
 std::vector<ssize_t> broadcast_shape(std::initializer_list<ssize_t> lhs,
                                      std::initializer_list<ssize_t> rhs);
 
+/// For "partial reduction", get the shape of the resulting shape of the array when preforming a
+/// reduction over an axis
+std::vector<ssize_t> partial_reduce_shape(const std::span<const ssize_t> array_shape,
+                                          const ssize_t axis);
+std::vector<ssize_t> partial_reduce_shape(std::initializer_list<ssize_t> input_shape,
+                                          const ssize_t axis);
+
+/// Gets the strides of a n-dimensional array assuming contiguous memory
+std::vector<ssize_t> as_contiguous_strides(const std::span<const ssize_t> shape);
```

Comment:
This duplicates the functionality of the existing [shape_to_strides()](https://github.com/dwavesystems/dwave-optimization/blob/a4e00cf2c24caec9eeab129a73c773da307af58e/dwave/optimization/include/dwave-optimization/array.hpp#L645) static method.

Reply:
I moved this behind compilation barrier in `mathematical.cpp`.